File: frontend/ui-architecture-playground/mvp-winforms/BoardApp/Presenters/BoardPresenter.cs
// ============================================================================
// BoardPresenter - ПРЕЗЕНТЕР главного экрана. "P" в MVP.
//
// Весь экран знает о мире только три вещи:
//   IBoardView  - чем показывать (интерфейс, не класс!)
//   IBoardStore - где данные (домен)
//   фабрики     - как открыть другие экраны (внедрены снаружи, см. Program.cs)
//
// Здесь НЕТ ни одного типа из System.Windows.Forms - и это не случайность:
// презентер можно запустить в консоли или в юнит-тесте (см. SmokeTest).
// Сравни: в MVC эту логику размазали бы по контроллерам + Razor-хелперам.
//
// Жизненный цикл: один презентер на всё время жизни окна (в отличие от веба,
// где контроллер рождается на каждый запрос). Поэтому он может хранить
// состояние (_currentBoardId) в обычном поле.
// ============================================================================

using BoardApp.Core;
using BoardApp.Views.Contracts;

namespace BoardApp.Presenters;

public sealed class BoardPresenter(
    IBoardView view,
    IBoardStore store,
    Action openUsersScreen)
{
    private Guid? _currentBoardId;

    // ------------------------------------------------------------------
    // Точка входа: подписаться на события view и сделать первый рендер.
    // Вызывается композиционным корнем (Program.cs) после связывания.
    // ------------------------------------------------------------------
    public void Run()
    {
        Subscribe();

        _currentBoardId = store.FirstBoard()?.Id;
        view.ResetFilters(AssigneeFilterOptions(store.Users), EpicOptions());
        Reload();
        view.ShowFlash(_currentBoardId is null
            ? "Создайте первую доску"
            : $"Открыта доска \"{CurrentBoard()?.Name}\"");
    }

    /// <summary>Внешний сигнал "данные изменились" (например, поработали с пользователями).</summary>
    public void ExternalRefresh() => Reload();

    // ==================================================================
    // ГЛАВНЫЙ РЕНДЕР. Полная перезагрузка данных во view после ЛЮБОЙ
    // мутации - прямой аналог PRG-перезагрузки страницы в веб-версиях:
    // сервер/презентер заново отдаёт весь экран, а не патчит кусочки.
    // ==================================================================
    private void Reload()
    {
        var board = CurrentBoard();
        var users = store.Users;

        // список досок в комбобоксе
        view.ShowBoards(
            store.Boards.Select(b => new BoardListItem(b.Id, b.Name)).ToList(),
            _currentBoardId);

        if (board is null)
        {
            // пустой мир: колонки без карточек, эпиков нет
            view.ShowColumns(Enum.GetValues<TaskState>()
                .Select(s => new ColumnVm(s, s.ToDisplay(), [])).ToList());
            view.ShowEpics([]);
            view.ShowTaskTable([]);
            return;
        }

        // --- фильтрация + сортировка (та же семантика, что в веб-версиях) ---
        var criteria = view.ReadFilterCriteria();
        var visible = FilterAndSort(board.Tasks, criteria);

        // канбан: группировка по колонкам с сохранением Order
        var columns = Enum.GetValues<TaskState>().Select(s => new ColumnVm(
            s,
            s.ToDisplay(),
            visible.Where(t => t.State == s).Select(ToCard).ToList())).ToList();
        view.ShowColumns(columns);

        // прогресс эпиков считается по ВСЕМ задачам доски (не по отфильтрованным)
        view.ShowEpics(board.Epics.Select(e => new EpicStatRow(
            e.Id,
            $"EPIC-{e.Number}",
            e.Title,
            board.Tasks.Count(t => t.EpicId == e.Id),
            board.Tasks.Count(t => t.EpicId == e.Id && t.State == TaskState.Done))).ToList());

        // вкладка "Список": те же отфильтрованные задачи плоским списком
        view.ShowTaskTable(visible.Select(t => new TaskRow(
            $"TASK-{t.Number}",
            t.Title,
            t.State.ToDisplay(),
            t.PriorityLevel.ToDisplay(),
            users.FirstOrDefault(u => u.Id == t.AssigneeId)?.Name ?? "—",
            t.Deadline?.ToString("dd.MM.yyyy") ?? "—")).ToList());
    }

    // ------------------------------------------------------------------
    // Фильтры и сортировка. Правила 1-в-1 как в вебе:
    //   null = "не фильтруем", Guid.Empty = спецзначение "без исполнителя/эпика".
    // ------------------------------------------------------------------
    private static List<TaskItem> FilterAndSort(IEnumerable<TaskItem> tasks, FilterCriteria c)
    {
        IEnumerable<TaskItem> q = tasks;

        if (c.AssigneeId is { } a)
            q = q.Where(t => a == FilterSpecial.None ? t.AssigneeId is null : t.AssigneeId == a);
        if (c.EpicId is { } e)
            q = q.Where(t => e == FilterSpecial.None ? t.EpicId is null : t.EpicId == e);
        if (!string.IsNullOrWhiteSpace(c.SearchText))
        {
            var needle = c.SearchText.Trim();
            q = q.Where(t => t.Title.Contains(needle, StringComparison.OrdinalIgnoreCase)
                          || t.Description.Contains(needle, StringComparison.OrdinalIgnoreCase));
        }

        return c.SortMode == "priority"
            ? q.OrderBy(t => PriorityRank(t.PriorityLevel)).ThenBy(t => t.Order).ToList()
            : q.OrderBy(t => t.Order).ToList();
    }

    private static int PriorityRank(Priority p) => p switch
    {
        Priority.High => 0, Priority.Medium => 1, _ => 2
    };

    // ------------------------------------------------------------------
    // Маппинг домен -> DTO для карточек. View получает готовые строки.
    // ------------------------------------------------------------------
    private TaskCardVm ToCard(TaskItem t)
    {
        var board = CurrentBoard()!;
        return new TaskCardVm(
            t.Id,
            $"TASK-{t.Number}",
            t.Title,
            t.Type.ToDisplay(),
            t.PriorityLevel.ToDisplay(),
            t.PriorityLevel == Priority.High,
            store.Users.FirstOrDefault(u => u.Id == t.AssigneeId)?.Name,
            t.Deadline,
            t.Deadline is { } d && d < DateOnly.FromDateTime(DateTime.Today) && t.State != TaskState.Done,
            t.EpicId is null ? null : $"EPIC-{board.Epics.First(e => e.Id == t.EpicId).Number}");
    }

    private Board? CurrentBoard() =>
        _currentBoardId is { } id ? store.FindBoard(id) : null;

    // Опции фильтра исполнителя: спецпункт "без исполнителя" + все пользователи.
    // Пункт "(все)" (фильтр не активен) во view добавляет сама - он означает null.
    private static IReadOnlyList<OptionVm> AssigneeFilterOptions(IReadOnlyList<BoardUser> users) =>
        new[] { new OptionVm(FilterSpecial.None, "Без исполнителя") }
            .Concat(users.Select(u => new OptionVm(u.Id, u.Name)))
            .ToList();

    private IReadOnlyList<OptionVm> EpicOptions() =>
        CurrentBoard()?.Epics
            .Select(e => new OptionVm(e.Id, $"EPIC-{e.Number} · {e.Title}"))
            .Prepend(new OptionVm(FilterSpecial.None, "Без эпика"))
            .ToList() ?? [];

    // ==================================================================
    // ОБРАБОТЧИКИ СОБЫТИЙ. Каждый: прочитать ввод -> мутировать домен ->
    // Reload() -> flash. Никакой логики во view.
    // ==================================================================

    private void Subscribe()
    {
        view.CreateBoardRequested += (_, _) => CreateBoard();
        view.RenameBoardRequested += (_, _) => RenameBoard();
        view.DeleteBoardRequested += (_, _) => DeleteBoard();
        view.SwitchBoardRequested += (_, e) => SwitchTo(e.Id);
        view.CreateTaskRequested += (_, state) => OpenTaskDialog(null, state);
        view.TaskOpenRequested += (_, e) => OpenTaskDialog(e.Id, null);
        view.TaskDeleteRequested += (_, e) => DeleteTask(e.Id);
        view.TaskMoved += (_, e) => MoveTask(e);
        view.ApplyFiltersRequested += (_, _) => Reload();
        view.ResetFiltersRequested += (_, _) =>
        {
            view.ResetFilters(AssigneeFilterOptions(store.Users), EpicOptions());
            Reload();
        };
        view.SeedEpicRequested += (_, _) => SeedEpic();
        view.SeedTasksRequested += (_, _) => SeedTasks();
        view.CreateEpicRequested += (_, _) => CreateEpic();
        view.EpicDeleteRequested += (_, e) => DeleteEpic(e.Id);
        view.ExportRequested += (_, _) => Export();
        view.ImportRequested += (_, _) => Import();
        view.ResetAllRequested += (_, _) => ResetAll();
        view.OpenUsersRequested += (_, _) => openUsersScreen();
    }

    private void CreateBoard()
    {
        var name = view.Prompt("Новая доска", "Название доски:");
        if (string.IsNullOrWhiteSpace(name)) return;              // отмена или пусто

        var board = store.CreateBoard(name.Trim());
        _currentBoardId = board.Id;                                // gherkin: сразу переключаемся
        Reload();
        view.ShowFlash($"Доска \"{board.Name}\" создана");
    }

    private void RenameBoard()
    {
        var board = CurrentBoard();
        if (board is null) return;
        var name = view.Prompt("Переименовать доску", "Новое название:", board.Name);
        if (string.IsNullOrWhiteSpace(name)) return;

        store.RenameBoard(board.Id, name.Trim());
        Reload();
        view.ShowFlash("Доска переименована");
    }

    private void DeleteBoard()
    {
        var board = CurrentBoard();
        if (board is null) return;
        if (!view.Confirm($"Удалить доску \"{board.Name}\" вместе со всеми задачами?")) return;

        store.DeleteBoard(board.Id);
        _currentBoardId = store.FirstBoard()?.Id;                  // gherkin: показать последнюю оставшуюся
        Reload();
        view.ShowFlash("Доска удалена");
    }

    private void SwitchTo(Guid id)
    {
        if (_currentBoardId == id) return;                         // событие при перезаполнении списка
        _currentBoardId = id;
        Reload();                                                  // gherkin: последняя открытая = текущая в окне
    }

    /// <summary>
    /// Диалог задачи: создание (taskId=null) или редактирование.
    /// Обратите внимание на цикл валидации: пустой заголовок -> ошибка -> диалог
    /// снова открыт. Решение о валидности принимает ПРЕЗЕНТЕР, а не форма.
    /// </summary>
    private void OpenTaskDialog(Guid? taskId, TaskState? defaultState)
    {
        var board = CurrentBoard();
        if (board is null) return;

        var existing = taskId is { } tid ? board.Tasks.FirstOrDefault(t => t.Id == tid) : null;

        // Фабрика диалога внедрена снаружи (Program.cs): презентер не знает,
        // что это WinForms Form, - только контракт ITaskEditView.
        var dlg = _taskEditFactory();

        dlg.DialogTitle = existing is null ? "Новая задача" : $"Редактирование TASK-{existing.Number}";
        dlg.FillOptions(
            store.Users.Select(u => new OptionVm(u.Id, u.Name)).ToList(),
            board.Epics.Select(e => new OptionVm(e.Id, $"EPIC-{e.Number} · {e.Title}")).ToList());
        dlg.DefaultState = defaultState ?? existing?.State ?? TaskState.ToDo;

        if (existing is not null)
        {
            dlg.Title = existing.Title;
            dlg.Description = existing.Description;
            dlg.AssigneeId = existing.AssigneeId;
            dlg.EpicId = existing.EpicId;
            dlg.State = existing.State;
            dlg.Type = existing.Type;
            dlg.Priority = existing.PriorityLevel;
            dlg.Deadline = existing.Deadline;
        }
        else
        {
            dlg.Title = ""; dlg.Description = "";
            dlg.AssigneeId = null; dlg.EpicId = null;
            dlg.Type = WorkItemType.Task; dlg.Priority = Priority.Medium; dlg.Deadline = null;
        }

        while (true)
        {
            if (!dlg.ShowModal()) return;                          // отмена

            // Диагностика «Сохранить ничего не делает» -> %TEMP%\kanban-errors\winforms.log:
            // видно каждую итерацию цикла валидации и финальное создание задачи.
            Program.LogTrace($"EditTask: ShowModal=OK, Title='{dlg.Title}'");

            if (string.IsNullOrWhiteSpace(dlg.Title))
            {
                dlg.ShowValidationError("Заголовок обязателен.");
                continue;
            }

            if (existing is null)
            {
                var created = store.AddTask(board.Id, new NewTask(
                    dlg.Title.Trim(), dlg.Description.Trim(),
                    dlg.AssigneeId, dlg.EpicId,
                    dlg.State, dlg.Type, dlg.Priority, dlg.Deadline));
                Reload();
                view.ShowFlash(created is null ? "Не удалось создать задачу" : $"TASK-{created.Number} создана");
            }
            else
            {
                store.UpdateTask(board.Id, existing.Id, t =>
                {
                    t.Title = dlg.Title.Trim();
                    t.Description = dlg.Description.Trim();
                    t.AssigneeId = dlg.AssigneeId;
                    t.EpicId = dlg.EpicId;
                    t.State = dlg.State;
                    t.Type = dlg.Type;
                    t.PriorityLevel = dlg.Priority;
                    t.Deadline = dlg.Deadline;
                });
                Reload();
                view.ShowFlash($"TASK-{existing.Number} сохранена");
            }
            return;
        }
    }

    // фабрика диалога задачи: внедряется из Program.cs через UseTaskEditDialog
    private Func<ITaskEditView> _taskEditFactory = () =>
        throw new InvalidOperationException("Фабрика диалогов не подключена (см. UseTaskEditDialog)");

    public BoardPresenter UseTaskEditDialog(Func<ITaskEditView> factory)
    {
        _taskEditFactory = factory;
        return this;
    }

    private void DeleteTask(Guid taskId)
    {
        var board = CurrentBoard();
        var task = board?.Tasks.FirstOrDefault(t => t.Id == taskId);
        if (task is null) return;
        if (!view.Confirm($"Удалить TASK-{task.Number} \"{task.Title}\"?")) return;

        store.DeleteTask(board!.Id, task.Id);
        Reload();
        view.ShowFlash($"TASK-{task.Number} удалена");
    }

    private void MoveTask(TaskMovedEventArgs e)
    {
        var board = CurrentBoard();
        if (board is null) return;
        store.MoveTask(board.Id, e.TaskId, e.TargetColumn, e.Index);   // инварианты Order - внутри стора
        Reload();                                                       // SSR-стиль: полный перерендер
    }

    private void SeedEpic()
    {
        var board = CurrentBoard();
        if (board is null) return;
        var epic = store.SeedTestEpic(board.Id);
        Reload();
        view.ShowFlash(epic is null ? "Не удалось создать эпик" : $"EPIC-{epic.Number} с тестовыми задачами добавлен");
    }

    private void SeedTasks()
    {
        var board = CurrentBoard();
        if (board is null) return;
        store.SeedRandomTasks(board.Id, 10);
        Reload();
        view.ShowFlash("Добавлено 10 случайных задач");
    }

    // gherkin Epic Management: создание эпика через промпт заголовка.
    private void CreateEpic()
    {
        var board = CurrentBoard();
        if (board is null) return;

        var title = view.Prompt("Новый эпик", "Название эпика:");
        if (string.IsNullOrWhiteSpace(title)) return;

        var epic = store.AddEpic(board.Id, title.Trim(), description: null);
        Reload();
        view.ShowFlash(epic is null ? "Не удалось создать эпик" : $"EPIC-{epic.Number} создан");
    }

    // gherkin: при удалении непустого эпика спрашиваем режим -
    // отвязать задачи или удалить каскадом. Решение показывает ДИАЛОГ view,
    // но интерпретирует его результат presenter.
    private void DeleteEpic(Guid epicId)
    {
        var board = CurrentBoard();
        var epic = board?.Epics.FirstOrDefault(e => e.Id == epicId);
        if (epic is null) return;

        var count = board!.Tasks.Count(t => t.EpicId == epicId);
        var mode = view.ChooseEpicDeleteMode($"EPIC-{epic.Number}", epic.Title, count);
        if (mode is null) return;                                  // передумал

        store.DeleteEpic(board.Id, epicId, mode.Value);
        Reload();
        view.ShowFlash(mode == EpicDeleteMode.CascadeDeleteTasks
            ? $"EPIC-{epic.Number} удалён вместе с задачами"
            : $"EPIC-{epic.Number} удалён, задачи остались");
    }

    private void Export()
    {
        var stamp = DateTime.Now.ToString("yyyyMMdd-HHmmss");
        if (view.SaveToFile($"kanban-export-{stamp}.json", store.ExportJson()))
            view.ShowFlash("Данные экспортированы в JSON");
    }

    private void Import()
    {
        var json = view.OpenJsonFile();
        if (json is null) return;
        try
        {
            store.ImportJson(json);
            _currentBoardId = store.FirstBoard()?.Id;
            view.ResetFilters(AssigneeFilterOptions(store.Users), EpicOptions());
            Reload();
            view.ShowFlash("Импорт выполнен");
        }
        catch (Exception ex)
        {
            view.ShowFlash($"Ошибка импорта: {ex.Message}");       // битый JSON - штатная ситуация
        }
    }

    private void ResetAll()
    {
        var word = view.AskConfirmWord("полного сброса всех данных");
        if (word != "СБРОС")                                       // точное слово, регистр важен
        {
            view.ShowFlash("Сброс отменён (нужно слово СБРОС)");
            return;
        }
        store.ResetAll();
        _currentBoardId = null;
        view.ResetFilters(AssigneeFilterOptions(store.Users), EpicOptions());
        Reload();
        view.ShowFlash("Все данные удалены");
    }
}

File: frontend/ui-architecture-playground/mvp-vanillajs/js/presenters/board-presenter.js
// ============================================================================
// BoardPresenter - "P" в MVP. JS-порт C#-презентера, логика 1:1.
//
// Знает только: view (объект с методами контракта), store (Model),
// фабрику диалога задачи. Ноль обращений к DOM - поэтому вся логика
// экрана проверяется автотестом с фейковой вьюхой (см. README).
//
// Отличие от веб-MVC: презентер ЖИВЁТ столько же, сколько страница,
// и держит состояние (_currentBoardId) в обычном поле - как в WinForms.
// ============================================================================

'use strict';

class BoardPresenter {
    constructor(view, store, { openUsersScreen, taskEditFactory }) {
        this.view = view;
        this.store = store;
        this.openUsersScreen = openUsersScreen;
        this.taskEditFactory = taskEditFactory ??
            (() => { throw new Error('Фабрика диалога задачи не подключена'); });
        this.currentBoardId = null;
    }

    /** Точка входа: подписаться на события view + первый рендер (аналог Run()). */
    run() {
        this.view.bindHandlers({
            createBoard: () => this.createBoard(),
            renameBoard: () => this.renameBoard(),
            deleteBoard: () => this.deleteBoard(),
            switchBoard: id => this.switchTo(id),
            createTask: state => this.openTaskDialog(null, state),
            openTask: id => this.openTaskDialog(id, null),
            deleteTask: id => this.deleteTask(id),
            taskMoved: (taskId, state, index) => this.moveTask(taskId, state, index),
            applyFilters: () => this.reload(),
            resetFilters: () => {
                this.view.resetFilters(assigneeFilterOptions(this.store.users), this.epicOptions());
                this.reload();
            },
            seedEpic: () => this.seedEpic(),
            seedTasks: () => this.seedTasks(),
            createEpic: () => this.createEpic(),
            deleteEpic: id => this.deleteEpic(id),
            exportData: () => this.exportData(),
            importData: text => this.importData(text),
            resetAll: () => this.resetAll(),
            openUsers: () => this.openUsersScreen()
        });

        this.currentBoardId = this.store.firstBoard()?.id ?? null;
        this.view.resetFilters(assigneeFilterOptions(this.store.users), this.epicOptions());
        this.reload();
        this.view.flash(this.currentBoardId
            ? `Открыта доска «${this.#board().name}»`
            : 'Создайте первую доску');
    }

    /** Внешний сигнал "данные изменились" (аналог ExternalRefresh). */
    externalRefresh() { this.reload(); }

    // ---------------- главный рендер (SSR-стиль: всё заново) ----------------

    reload() {
        const board = this.#board();
        this.view.renderBoards(
            this.store.boards.map(b => ({ id: b.id, name: b.name })),
            this.currentBoardId);

        if (!board) {
            this.view.renderColumns(emptyColumns());
            this.view.renderEpics([]);
            this.view.showTable([]);
            return;
        }

        const criteria = this.view.readFilterCriteria();
        const visible = filterAndSort(board.tasks, criteria);

        this.view.renderColumns(Object.values(TaskState).map(state => ({
            state,
            title: stateTitle(state),
            cards: visible.filter(t => t.state === state).map(t => this.toCard(board, t))
        })));

        // прогресс эпиков - по ВСЕМ задачам доски, не по отфильтрованным
        this.view.renderEpics(board.epics.map(e => ({
            id: e.id,
            key: `EPIC-${e.number}`,
            title: e.title,
            total: board.tasks.filter(t => t.epicId === e.id).length,
            done: board.tasks.filter(t => t.epicId === e.id && t.state === TaskState.Done).length
        })));

        this.view.showTable(visible.map(t => ({
            key: `TASK-${t.number}`,
            title: t.title,
            state: stateTitle(t.state),
            priority: priorityTitle(t.priority),
            assignee: this.store.users.find(u => u.id === t.assigneeId)?.name ?? '—',
            deadline: t.deadline ?? '—'
        })));
    }

    toCard(board, t) {
        const today = new Date().toISOString().slice(0, 10);
        return {
            id: t.id,
            key: `TASK-${t.number}`,
            title: t.title,
            typeName: typeTitle(t.type),
            priorityName: priorityTitle(t.priority),
            priorityClass: t.priority,                              // high|medium|low -> CSS
            assignee: this.store.users.find(u => u.id === t.assigneeId)?.name ?? null,
            deadlineText: t.deadline ? `⏰ ${formatDate(t.deadline)}` : null,
            overdue: !!t.deadline && t.deadline < today && t.state !== TaskState.Done,
            epicKey: t.epicId ? `EPIC-${board.epics.find(e => e.id === t.epicId)?.number}` : null
        };
    }

    #board() { return this.currentBoardId ? this.store.findBoard(this.currentBoardId) : null; }

    epicOptions() {
        const board = this.#board();
        if (!board) return [];
        return [{ id: FILTER_NONE, label: 'Без эпика' },
                ...board.epics.map(e => ({ id: e.id, label: `EPIC-${e.number} · ${e.title}` }))];
    }

    // ---------------- обработчики досок ----------------

    async createBoard() {
        const name = await this.view.prompt('Новая доска', 'Название доски:');
        if (!name?.trim()) return;
        const board = this.store.createBoard(name.trim());
        this.currentBoardId = board.id;                             // gherkin: сразу переключаемся
        this.reload();
        this.view.flash(`Доска «${board.name}» создана`);
    }

    async renameBoard() {
        const board = this.#board();
        if (!board) return;
        const name = await this.view.prompt('Переименовать доску', 'Новое название:', board.name);
        if (!name?.trim()) return;
        this.store.renameBoard(board.id, name.trim());
        this.reload();
        this.view.flash('Доска переименована');
    }

    async deleteBoard() {
        const board = this.#board();
        if (!board) return;
        if (!(await this.view.confirmBox(`Удалить доску «${board.name}» вместе со всеми задачами?`))) return;
        this.store.deleteBoard(board.id);
        this.currentBoardId = this.store.firstBoard()?.id ?? null;  // gherkin: показать оставшуюся
        this.reload();
        this.view.flash('Доска удалена');
    }

    switchTo(id) {
        if (this.currentBoardId === id) return;                     // эхо перерисовки списка
        this.currentBoardId = id;
        this.reload();
    }

    // ---------------- диалог задачи (цикл валидации как в C#) ----------------

    async openTaskDialog(taskId, defaultState) {
        const board = this.#board();
        if (!board) return;

        const existing = taskId ? board.tasks.find(t => t.id === taskId) : null;
        const data = await this.taskEditFactory()({
            title: existing ? `Редактирование TASK-${existing.number}` : 'Новая задача',
            assignees: this.store.users.map(u => ({ id: u.id, label: u.name })),
            epics: board.epics.map(e => ({ id: e.id, label: `EPIC-${e.number} · ${e.title}` })),
            task: existing ? {
                title: existing.title, description: existing.description,
                assigneeId: existing.assigneeId, epicId: existing.epicId,
                state: existing.state, type: existing.type,
                priority: existing.priority, deadline: existing.deadline
            } : null,
            defaultState: defaultState ?? TaskState.ToDo
        });
        if (!data) return;                                          // отмена

        if (!existing) {
            const created = this.store.addTask(board.id, data);
            this.reload();
            this.view.flash(created ? `TASK-${created.number} создана` : 'Не удалось создать задачу');
        } else {
            this.store.updateTask(board.id, existing.id, t => Object.assign(t, data));
            this.reload();
            this.view.flash(`TASK-${existing.number} сохранена`);
        }
    }

    async deleteTask(taskId) {
        const board = this.#board();
        const task = board?.tasks.find(t => t.id === taskId);
        if (!task) return;
        if (!(await this.view.confirmBox(`Удалить TASK-${task.number} «${task.title}»?`))) return;
        this.store.deleteTask(board.id, task.id);
        this.reload();
        this.view.flash(`TASK-${task.number} удалена`);
    }

    moveTask(taskId, targetState, index) {
        const board = this.#board();
        if (!board) return;
        this.store.moveTask(board.id, taskId, targetState, index);  // инварианты Order внутри стора
        this.reload();
    }

    // ---------------- эпики ----------------

    async createEpic() {
        const board = this.#board();
        if (!board) return;
        const title = await this.view.prompt('Новый эпик', 'Название эпика:');
        if (!title?.trim()) return;
        const epic = this.store.addEpic(board.id, title.trim(), '');
        this.reload();
        this.view.flash(epic ? `EPIC-${epic.number} создан` : 'Не удалось создать эпик');
    }

    async deleteEpic(epicId) {
        const board = this.#board();
        const epic = board?.epics.find(e => e.id === epicId);
        if (!epic) return;

        const count = board.tasks.filter(t => t.epicId === epicId).length;
        const mode = await this.view.chooseEpicDeleteMode(`EPIC-${epic.number}`, epic.title, count);
        if (!mode) return;
        this.store.deleteEpic(board.id, epicId, mode);
        this.reload();
        this.view.flash(mode === 'cascade'
            ? `EPIC-${epic.number} удалён вместе с задачами`
            : `EPIC-${epic.number} удалён, задачи остались`);
    }

    async seedEpic() {
        const board = this.#board();
        if (!board) return;
        const epic = this.store.seedTestEpic(board.id);
        this.reload();
        this.view.flash(epic ? `EPIC-${epic.number} с тестовыми задачами добавлен` : 'Не удалось создать эпик');
    }

    seedTasks() {
        const board = this.#board();
        if (!board) return;
        this.store.seedRandomTasks(board.id, 10);
        this.reload();
        this.view.flash('Добавлено 10 случайных задач');
    }

    // ---------------- персистентность ----------------

    exportData() {
        const stamp = new Date().toISOString().slice(0, 19).replace(/[:T]/g, '-');
        this.view.saveFile(`kanban-export-${stamp}.json`, this.store.exportJson());
        this.view.flash('Данные экспортированы в JSON');
    }

    async importData(text) {
        try {
            this.store.importJson(text);
            this.currentBoardId = this.store.firstBoard()?.id ?? null;
            this.view.resetFilters(assigneeFilterOptions(this.store.users), this.epicOptions());
            this.reload();
            this.view.flash('Импорт выполнен');
        } catch (ex) {
            this.view.flash(`Ошибка импорта: ${ex.message}`);       // битый JSON - штатная ситуация
        }
    }

    async resetAll() {
        const word = await this.view.askConfirmWord('полного сброса всех данных');
        if (word !== 'СБРОС') {                                     // точное слово, регистр важен
            this.view.flash('Сброс отменён (нужно слово СБРОС)');
            return;
        }
        this.store.resetAll();
        this.currentBoardId = null;
        this.view.resetFilters(assigneeFilterOptions(this.store.users), []);
        this.reload();
        this.view.flash('Все данные удалены');
    }
}

// ---------------- общие правила фильтрации (те же, что в вебе) ----------------
// Спецзначение FILTER_NONE ("без исполнителя/эпика") объявлено в models.js.

function assigneeFilterOptions(users) {
    return [{ id: FILTER_NONE, label: 'Без исполнителя' },
            ...users.map(u => ({ id: u.id, label: u.name }))];
}

function filterAndSort(tasks, c) {
    let q = [...tasks];
    if (c.assigneeId)
        q = q.filter(t => c.assigneeId === FILTER_NONE ? !t.assigneeId : t.assigneeId === c.assigneeId);
    if (c.epicId)
        q = q.filter(t => c.epicId === FILTER_NONE ? !t.epicId : t.epicId === c.epicId);
    if (c.searchText)
        q = q.filter(t => (t.title + ' ' + t.description)
            .toLowerCase().includes(c.searchText.toLowerCase()));

    const rank = p => p === Priority.High ? 0 : p === Priority.Medium ? 1 : 2;
    return c.sortMode === 'priority'
        ? q.sort((a, z) => rank(a.priority) - rank(z.priority) || a.order - z.order)
        : q.sort((a, z) => a.order - z.order);
}

function emptyColumns() {
    return Object.values(TaskState).map(state => ({ state, title: stateTitle(state), cards: [] }));
}

function formatDate(iso) {
    const [y, m, d] = iso.split('-');
    return `${d}.${m}.${y}`;
}
